biahub/track.py

```diff
@@ -0,0 +1,993 @@
+
+    # Prepare SLURM arguments
+    slurm_args = {
+        "slurm_job_name": "tracking",
+        "slurm_mem_per_cpu": f"{gb_ram_per_cpu}G",
+        "slurm_cpus_per_task": num_cpus,
+        "slurm_array_parallelism": 100,  # process up to 100 positions at a time
+        "slurm_time": 60,
+        "slurm_partition": "preempted",
+        "slurm_num_gpus": 1,
+        "slurm_use_srun": False,
+    }
+
+    # Override defaults if sbatch_filepath is provided
+    if sbatch_filepath:
+        slurm_args.update(sbatch_to_submitit(sbatch_filepath))
+
+    # Run locally or submit to SLURM
+    if local:
+        cluster = "local"
+    else:
+        cluster = "slurm"
+
+    # Prepare and submit jobs
+    click.echo(f"Preparing jobs: {slurm_args}")
+    slurm_out_path = output_dirpath.parent / "slurm_output"
+    executor = submitit.AutoExecutor(folder=slurm_out_path, cluster=cluster)
+    executor.update_parameters(**slurm_args)
+
+    click.echo('Submitting SLURM jobs...')
+    jobs = []
+
+    with executor.batch():
+        for input_position_path in input_position_dirpaths:
+            position_key = input_position_path.parts[-3:]
+            job = executor.submit(
+                track_one_position,
+                position_key=position_key,
+                output_dirpath=output_dirpath,
+                tracking_config=tracking_cfg,
+                input_images=settings.input_images,
+                blank_frames_path=blank_frames_path,
+                z_slices=z_slices,
+                scale=track_scale,
+            )
+
+            jobs.append(job)
+
+    job_ids = [job.job_id for job in jobs]  # Access job IDs after batch submission
+
+    log_path = Path(slurm_out_path / "submitit_jobs_ids.log")
+    with log_path.open("w") as log_file:
+        log_file.write("\n".join(job_ids))
+
+
+@click.command("track")
+@input_position_dirpaths()
+@output_dirpath()
+@config_filepath()
+@sbatch_filepath()
+@local()
+@click.option(
+    "--blank_frames_path",
+    "-f",
+    required=False,
+    default=None,
+    type=str,
+    help="Path to blank frame CSV file, if available.",
+)
+def track_cli(
+    input_position_dirpaths: List[str],
+    output_dirpath: str,
+    config_filepath: str,
+    sbatch_filepath: str = None,
+    local: bool = None,
+    blank_frames_path: str = None,
+) -> None:
+
+    """
+    Track objects in 2D or 3D time-lapse microscopy data using configurable preprocessing.
+
+    This command applies preprocessing, handles optional blank frame filling, and performs
+    object tracking on each position using the Ultrack library. Compatible with any image
+    modality as long as it produces 'foreground' and 'contour' inputs.
+
+    Example usage:
+
+    biahub track -i virtual_staining.zarr/*/*/* -o output.zarr -c config_tracking.yml -f blank_frames.csv
+
+    """
+
+    track(
+        output_dirpath=output_dirpath,
+        config_filepath=config_filepath,
+        input_position_dirpaths=input_position_dirpaths,
```

Comment:
How do the `input_position_dirpaths` here work together with the image paths given in the tracking config file? From what I understand, the images provided in the config file are used to compute contours which are used to perform the tracking. Once we have tracks, they are overlayed onto the data in `input_position_dirpaths`. Is that right? Is the data pass with `input_position_dirpaths` modified in any way? Should the output of the tracking CLI be a CSV file that could be associated with any zarr store?

Reply:
I see that tracking generated a CSV file and a zarr file with object labels. I'm guessing these are added as channels to the data in `input_position_dirpaths`? I suggest that tracking output only a zarr with the labels - these can then be concatenated to any other zarr store. This would lead to less data duplication